### PR TITLE
Feat/updater conf

### PR DESCRIPTION
**Pull request by amir-khassaia**
- Feel free to review commit by commit to see the "journey"
- There are several things happening here: the introduction of updaterconf package that builds upon silver service config management, splitting of monolithic updater into CLI and reusable package, some clean ups

### CODE REVIEW

File: updater/update/update.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+type Operation struct {
+	Action string
+	Args   []string
+}
+
+func Check(url string, currentVer string) (*UpgradeInfo, error) {
+	client := &http.Client{}
+	req, err := http.NewRequest("GET", url+"?version="+currentVer, nil)
```
Comment (tinhtruong) [inline updater/update/update.go:30]: maybe encode query param the currentVer to be on the safe?

File: updater/update/update.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("User-Agent", "Update Check")
```
Comment (tinhtruong) [inline updater/update/update.go:34]: NIT: Should we add some kind of version to help us identify the updater in the future?

File: updater/update/update.go (hunk 1 above)
Comment (amir-khassaia) [reply updater/update/update.go:30]: This is unchanged really to how it was for all these years, but suppose can do with extra safety

File: updater/update/update.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode == http.StatusNotModified {
+		return nil, nil
+	}
+
+	if res.StatusCode >= http.StatusBadRequest {
+		return nil, fmt.Errorf("Got an error from the update url: %s", res.Status)
```
Comment (tinhtruong) [inline updater/update/update.go:48]: Include the StatusCode as well?

File: updater/update/update.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+		case "batchrename", "batch-rename":
+			fn = BatchRenameOp
+		case "move", "mv":
+			fn = MoveOp
+		case "copy", "cp":
+			fn = CopyOp
+		case "remove", "rm", "del", "delete":
+			fn = RemoveOp
+		default:
+			msg := fmt.Sprintf("Invalid operation action: '%s'", action)
```
Comment (tinhtruong) [inline updater/update/update.go:102]: '%s' -> %q ?

File: updater/update/update.go (hunk 2 above)
Comment (amir-khassaia) [reply updater/update/update.go:34]: yep I do agree the version could be here...but not trying to solve every problem, just splitting out the updater out of gigantic single file CLI

File: service/config/config.go [hunk 5]
```diff
@@ -100,8 +101,8 @@ type ReplacementVars struct {
 
 // LoadConfig parses config.
 func LoadConfig(path string, vars ReplacementVars) (conf *Config, err error) {
-	if _, err = os.Stat(path); os.IsNotExist(err) {
-		return nil, fmt.Errorf("The configuration file does not exist. Please place the file here: %s", path)
+	if !osutils.FileExists(path) {
+		return nil, fmt.Errorf("The conf file does not exist. Please configuration here: %s", path)
```
Comment (tinhtruong) [inline service/config/config.go:105]: return nil, fmt.Errorf("The conf file does not exist. Please configuration here: %s", path)
->
return nil, fmt.Errorf("The conf file does not exist. Please put the configuration file here: %s", path)

File: updater/update/update.go (hunk 4 above)
Comment (amir-khassaia) [reply updater/update/update.go:102]: Yep, true its silly its doing this

File: service/updaterconf/updaterconf.go [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+			return true
+		}
+	}
+	return false
+}
+
+func (u *UpdaterConf) getTasksFromServiceConfig() ([]config.Task, error) {
+	conf, err := config.LoadConfigNoReplacements(u.getServiceConfigPath())
+	if err != nil {
+		return []config.Task{}, err
```
Comment (tinhtruong) [inline service/updaterconf/updaterconf.go:220]: Can we **return nil, err** here?

Comment (amir-khassaia) [reply service/updaterconf/updaterconf.go:220]: Will do.

File: service/config/config.go (hunk 5 above)
Comment (amir-khassaia) [reply service/config/config.go:105]: will do

File: updater/update/update.go (hunk 3 above)
Comment (amir-khassaia) [reply updater/update/update.go]: done

File: updater/update/update.go (hunk 1 above)
Comment (amir-khassaia) [reply updater/update/update.go]: done

File: service/updaterconf/updaterconf.go [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+	if err != nil {
+		fmt.Printf("disableUpdates could not load config file: %v\n", err)
+		return err
+	}
+
+	_, ok := conf[startupTasksKey]
+	if ok {
+		delete(conf, startupTasksKey)
+	}
+	_, ok = conf[scheduledTasksKey]
```
Comment (pc-rowan) [inline service/updaterconf/updaterconf.go:180]: This will remove all tasks, but is there a chance there are tasks which is not the updater?

Comment (amir-khassaia) [reply service/updaterconf/updaterconf.go:180]: Yep you're right, whilst its unlikely this to be the case, in the context of silver this is a general purpose feature that needs to be aware of this distinction

File: lib/osutils/file.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+	return d.Close()
+}
+
+// WriteFileString writes a string to file denoted by filename with specified permissions.
+func WriteFileString(filename string, data string, perm os.FileMode) error {
+	f, err := os.OpenFile(filename, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, perm)
+	if err != nil {
+		return err
+	}
+	_, err = f.WriteString(data)
```
Comment (McCraeAndrew) [inline lib/osutils/file.go:44]: we will only return this error if there is no error closing the file?

Comment (amir-khassaia) [reply lib/osutils/file.go:44]: Yep, will add extra if

File: service/updaterconf/updaterconf.go (hunk 7 above)
Comment (amir-khassaia) [reply service/updaterconf/updaterconf.go]: updated with all the acrobatics required to deal with filtering at the map level 😩

**Review by McCraeAndrew — APPROVED** [review]

File: updater/update/updateops.go [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+			return err
+		}
+	}
+	fmt.Printf("Renamed %d files.\n", renameCnt)
+	return nil
+}
+
+func MoveOp(args []string) error {
+	if len(args) != 2 {
+		return errors.New("Invalid copy operation format - two args expected.")
```
[the remark this answers is not in the document]
Comment (amir-khassaia) [reply updater/update/updateops.go:92]: I didnt want to touch those, they are all capitalized in entire repo

File: service/updaterconf/updaterconf.go [hunk 10]
```diff
@@ -196,6 +195,30 @@ func (u *UpdaterConf) DisableAutoUpdate() error {
...
+	for i := 0; i < len(tasks); i++ {
+		task, ok := tasks[i].(map[string]interface{})
+		if !ok {
+			continue
+		}
+		path, ok := task[pathKey].(string)
+		if !ok {
+			continue
+		}
+		if strings.Contains(path, u.updaterFilename) {
```
Comment (pc-rowan) [inline service/updaterconf/updaterconf.go:212]: Is updaterFilename configurable somewhere? I may have missed it.

Comment (amir-khassaia) [reply service/updaterconf/updaterconf.go:212]: yes part of call to
```
 func Create(
	silverDir string,
	silverConfigFilename string,
	updaterFilename string,
)
```

File: updater/update/update.go [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode == http.StatusNotModified {
+		return nil, nil
+	}
+
+	if res.StatusCode >= http.StatusBadRequest {
+		return nil, fmt.Errorf("Got an error from the update url: %d (%s) ", res.StatusCode, res.Status)
```
[the remark this answers is not in the document]
Comment (amir-khassaia) [reply updater/update/update.go:49]: I think this needs to be a separate PR to do the whole repo, I'm just moving this around and its all like this

File: lib/osutils/file.go [hunk 12]
```diff
@@ -0,0 +1,60 @@
+/*
```
Comment (pc-rowan) [inline lib/osutils/file.go:1]: Do we need tests for file backup, remove tasks etc?

Comment (amir-khassaia) [reply lib/osutils/file.go:1]: Ok I will add some

**Review by pc-rowan — APPROVED** [review]

**Pull request merged**